From f7be56ef9dee90818b787fa71067cad70d556415 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Mon, 24 Jan 2022 14:36:01 +0200
Subject: [PATCH] [SW-58071] gaudi3: add status update support

This patch adds device specific status update support for gaudi3
device.

Change-Id: I542cbb2bbfc26b218ae8c1d53312052645912961
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index 25d875c..342d958 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -28,13 +28,17 @@
 #define ARCBOOT_BRINGUP_ERROR		-1
 #define ARCBOOT_BRINGUP_OK		1
 
+#define ARCBOOT_STATUS_REG_SZ_BITS		32
 __weak void arcboot_rom_status_update(enum arcboot_rom_progress progress)
 {
-	uint32_t reg;
+	uint32_t reg, offset;
 
-	reg = regmap_readl(mmSW_STATUS_0);
+	offset = progress / ARCBOOT_STATUS_REG_SZ_BITS;
+	progress = progress % ARCBOOT_STATUS_REG_SZ_BITS;
+
+	reg = regmap_readl((mmSW_STATUS_0 + offset * 4));
 	reg |= BIT(progress);
-	regmap_writel(reg, mmSW_STATUS_0);
+	regmap_writel(reg, (mmSW_STATUS_0 + offset * 4));
 }
 
 static void arcboot_rom_status_update_bringup(enum arcboot_rom_progress progress,
diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 3a2486c..dad125a 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -19,7 +19,6 @@
 #include "spi_flash.h"
 #include "boot_desc.h"
 
-#define ARCBOOT_STATUS_REG_SZ_BITS	32
 #define LOW_DWORD(x)			(x & 0xFFFFFFFF)
 #define ICCM_OFFS_MASK			0x1FFFF
 #define SP_SRAM_OFFS_MASK		0xFFFF
@@ -35,18 +34,6 @@
 #define IATU_INBOUND_LOW_OFFSET_ADDR		0x0
 #define IATU_INBOUND_HI_OFFSET_ADDR		0xC0080000
 #define DBI_ENABLE				0x300000
-
-void arcboot_rom_status_update(enum arcboot_rom_progress progress)
-{
-	uint32_t reg, offset;
-
-	offset = progress / ARCBOOT_STATUS_REG_SZ_BITS;
-	progress = progress % ARCBOOT_STATUS_REG_SZ_BITS;
-
-	reg = regmap_readl((mmPSOC_SECURITY_SW_STATUS_0 + offset * 4));
-	reg |= BIT(progress);
-	regmap_writel(reg, (mmPSOC_SECURITY_SW_STATUS_0 + offset  * 4));
-}
 
 uint32_t get_mem_cfg_mask(void)
 {
diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 978c9fb..d142243 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -59,4 +59,6 @@
 
 	/* clear SW Status registers */
 	arcboot_rom_clear_sw_status();
+
+	arcboot_rom_status_update(PROGRESS_INIT);
 }
diff --git a/arch/greco/arc_elf32/common.c b/arch/greco/arc_elf32/common.c
index 6a7b738..2eb1a20 100644
--- a/arch/greco/arc_elf32/common.c
+++ b/arch/greco/arc_elf32/common.c
@@ -174,3 +174,11 @@
 	regmap_writel(write_en_16MB, mmPSOC_SECURITY_SPI_WRITE_LIMIT);
 }
 
+void arcboot_rom_status_update(enum arcboot_rom_progress progress)
+{
+	uint32_t reg;
+
+	reg = regmap_readl(mmSW_STATUS_0);
+	reg |= BIT(progress);
+	regmap_writel(reg, mmSW_STATUS_0);
+}
